File: examples/typescript/desktop_automation.ts
/**
 * EdgeBox TypeScript Desktop Automation Example
 *
 * Demonstrates using EdgeBox's desktop (Computer Use) capabilities
 * to control the GUI environment - take screenshots, move the mouse,
 * type text, and interact with desktop applications.
 *
 * Prerequisites:
 *   1. EdgeBox is running with Docker Desktop
 *   2. GUI Tools are ENABLED in EdgeBox settings
 *   3. Install dependencies: npm install
 *
 * Usage:
 *   npx tsx desktop_automation.ts
 */

import { Client } from "@modelcontextprotocol/sdk/client/index.js";
import { StreamableHTTPClientTransport } from "@modelcontextprotocol/sdk/client/streamableHttp.js";
import { SSEClientTransport } from "@modelcontextprotocol/sdk/client/sse.js";

const EDGEBOX_MCP_URL = "http://localhost:8888/mcp";

async function createClient(): Promise<Client> {
  const client = new Client(
    { name: "edgebox-ts-desktop", version: "1.0.0" },
    { capabilities: {} }
  );

  try {
    const transport = new StreamableHTTPClientTransport(
      new URL(EDGEBOX_MCP_URL)
    );
    await client.connect(transport);
  } catch {
    const sseTransport = new SSEClientTransport(
      new URL("http://localhost:8888/sse")
    );
    await client.connect(sseTransport);
  }

  return client;
}

async function main() {
  const client = await createClient();

  try {
    console.log("=".repeat(60));
    console.log("EdgeBox Desktop Automation Example");
    console.log("=".repeat(60));

    // --------------------------------------------------------
    // 1. Take a screenshot of the desktop
    // --------------------------------------------------------
    console.log("\n1. Taking a screenshot of the desktop...");
    const screenshot = await client.callTool({
      name: "desktop_screenshot",
      arguments: {},
    });
    console.log("  Screenshot captured:", screenshot.content);

    // --------------------------------------------------------
    // 2. List all open windows
    // --------------------------------------------------------
    console.log("\n2. Listing open windows...");
    const windows = await client.callTool({
      name: "desktop_get_windows",
      arguments: {},
    });
    console.log("  Windows:", windows.content);

    // --------------------------------------------------------
    // 3. Launch an application
    // --------------------------------------------------------
    console.log("\n3. Launching a text editor...");
    const launch = await client.callTool({
      name: "desktop_launch_app",
      arguments: { app_name: "xterm" },
    });
    console.log("  Launch result:", launch.content);

    // Wait for the application to open
    await client.callTool({
      name: "desktop_wait",
      arguments: { ms: 2000 },
    });

    // --------------------------------------------------------
    // 4. Type text using the keyboard
    // --------------------------------------------------------
    console.log("\n4. Typing text...");
    const typed = await client.callTool({
      name: "desktop_keyboard_type",
      arguments: { text: "echo 'Hello from EdgeBox Desktop Automation!'" },
    });
    console.log("  Typed text:", typed.content);

    // Press Enter to execute
    const enter = await client.callTool({
      name: "desktop_keyboard_press",
      arguments: { key: "Return" },
    });
    console.log("  Pressed Enter:", enter.content);

    // Wait for command to execute
    await client.callTool({
      name: "desktop_wait",
      arguments: { ms: 1000 },
    });

    // --------------------------------------------------------
    // 5. Take a final screenshot to see results
    // --------------------------------------------------------
    console.log("\n5. Taking final screenshot...");
    const finalScreenshot = await client.callTool({
      name: "desktop_screenshot",
      arguments: {},
    });
    console.log("  Final screenshot captured:", finalScreenshot.content);

    // --------------------------------------------------------
    // 6. Mouse operations
    // --------------------------------------------------------
    console.log("\n6. Mouse operations...");
    const move = await client.callTool({
      name: "desktop_mouse_move",
      arguments: { x: 100, y: 100 },
    });
    console.log("  Moved mouse to (100, 100):", move.content);

    const click = await client.callTool({
      name: "desktop_mouse_click",
      arguments: { x: 100, y: 100, button: "left" },
    });
    console.log("  Clicked at (100, 100):", click.content);

    console.log(`\n${"=".repeat(60)}`);
    console.log("Desktop automation example completed!");
    console.log("=".repeat(60));
  } finally {
    await client.close();
  }
}

main().catch(console.error);

File: examples/python/desktop_automation.py
"""
EdgeBox Python Desktop Automation Example

Demonstrates using EdgeBox's desktop (Computer Use) capabilities
to control the GUI environment - take screenshots, move the mouse,
type text, and interact with desktop applications.

Prerequisites:
  1. EdgeBox is running with Docker Desktop
  2. GUI Tools are ENABLED in EdgeBox settings
  3. Install dependencies: pip install -r requirements.txt

Usage:
  python desktop_automation.py
"""

import asyncio
from fastmcp import Client


EDGEBOX_MCP_URL = "http://localhost:8888/mcp"


async def main():
    client = Client(EDGEBOX_MCP_URL)

    async with client:
        print("=" * 60)
        print("EdgeBox Desktop Automation Example")
        print("=" * 60)

        # --------------------------------------------------------
        # 1. Take a screenshot of the desktop
        # --------------------------------------------------------
        print("\n1. Taking a screenshot of the desktop...")
        result = await client.call_tool("desktop_screenshot", {})
        print(f"  Screenshot captured: {result}")

        # --------------------------------------------------------
        # 2. List all open windows
        # --------------------------------------------------------
        print("\n2. Listing open windows...")
        result = await client.call_tool("desktop_get_windows", {})
        print(f"  Windows: {result}")

        # --------------------------------------------------------
        # 3. Launch an application
        # --------------------------------------------------------
        print("\n3. Launching a text editor...")
        result = await client.call_tool(
            "desktop_launch_app", {"app_name": "xterm"}
        )
        print(f"  Launch result: {result}")

        # Wait for the application to open
        await client.call_tool("desktop_wait", {"ms": 2000})

        # --------------------------------------------------------
        # 4. Type text using the keyboard
        # --------------------------------------------------------
        print("\n4. Typing text...")
        result = await client.call_tool(
            "desktop_keyboard_type",
            {"text": "echo 'Hello from EdgeBox Desktop Automation!'"},
        )
        print(f"  Typed text: {result}")

        # Press Enter to execute
        result = await client.call_tool(
            "desktop_keyboard_press", {"key": "Return"}
        )
        print(f"  Pressed Enter: {result}")

        # Wait for command to execute
        await client.call_tool("desktop_wait", {"ms": 1000})

        # --------------------------------------------------------
        # 5. Take a final screenshot to see results
        # --------------------------------------------------------
        print("\n5. Taking final screenshot...")
        result = await client.call_tool("desktop_screenshot", {})
        print(f"  Final screenshot captured: {result}")

        # --------------------------------------------------------
        # 6. Mouse operations
        # --------------------------------------------------------
        print("\n6. Mouse operations...")
        result = await client.call_tool(
            "desktop_mouse_move", {"x": 100, "y": 100}
        )
        print(f"  Moved mouse to (100, 100): {result}")

        result = await client.call_tool(
            "desktop_mouse_click", {"x": 100, "y": 100, "button": "left"}
        )
        print(f"  Clicked at (100, 100): {result}")

        print("\n" + "=" * 60)
        print("Desktop automation example completed!")
        print("=" * 60)


if __name__ == "__main__":
    asyncio.run(main())
